### PR TITLE
RavenDB-14325 Allow multiple tombstone in different collections

### PR DESCRIPTION
**Pull request by karmeli87**
So far we can't have two tombstones in different collections and we need for the cleaner to claim the old one before we can generate a new tombstone.
This is a problem because if you have an etl, external replication to a non-existing/broken destination the old tombstone will be never claimed and breaks replication for ever.

The idea here is to add a suffix to the tombstone ID if one is already exists (maybe we should always added the suffix?).
This way we don't have conflict. Indexing, replication and all other feature shouldn't behave differently, since strip the suffix when materializing the tombstone.

### CODE REVIEW

File: src/Raven.Server/Documents/DocumentsStorage.cs [hunk 1]
```diff
@@ -1207,6 +1208,7 @@ private static Tombstone TableValueToTombstone(JsonOperationContext context, ref
                 result.Collection = TableValueToId(context, (int)TombstoneTable.Collection, ref tvr);
                 result.Flags = TableValueToFlags((int)TombstoneTable.Flags, ref tvr);
                 result.LastModified = TableValueToDateTime((int)TombstoneTable.LastModified, ref tvr);
+                result.LowerId = UnwrapLowerIdIfNeeded(context, result.LowerId);
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/DocumentsStorage.cs:1224]: `result.LowerId` is `LazyStringValue`. Although here we force to materialize it as `string` because that's the argument of `UnwrapLowerIdIfNeeded`.

Please also see my comment below

File: src/Raven.Server/Documents/DocumentsStorage.cs [hunk 2]
```diff
@@ -1556,14 +1560,44 @@ public void EnsureLastEtagIsPersisted(DocumentsOperationContext context, long do
...
+            
+            return newLowerId;
+        }
+
+        private const long ConflictedTombstoneIdMarkerLong = 20L;
+        private static readonly string ConflictedTombstoneIdMarkerString = Encoding.UTF8.GetString(BitConverter.GetBytes(ConflictedTombstoneIdMarkerLong));
+
+        private static LazyStringValue UnwrapLowerIdIfNeeded(JsonOperationContext context, string lowerId)
+        {
+            var index = lowerId.IndexOf(ConflictedTombstoneIdMarkerString, StringComparison.OrdinalIgnoreCase);
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/DocumentsStorage.cs]: We `lowerId` be `LSV` instead of string? 

Then instead of calling `IndexOf` we could check if at position `lowerId.Size - 2*sizeof(long)` we have `ConflictedTombstoneIdMarkerString`?

File: src/Raven.Server/Documents/DocumentsStorage.cs [hunk 3]
```diff
@@ -1556,14 +1560,44 @@ public void EnsureLastEtagIsPersisted(DocumentsOperationContext context, long do
...
+            var length = lowerId.Content.Length;
+            Slice.From(context.Allocator, lowerId.Content.Ptr, length + sizeof(long) * 2, out var newLowerId); // TODO: dispose?
+
+            *(long*)(newLowerId.Content.Ptr + length) = ConflictedTombstoneIdMarkerLong;
+            *(long*)(newLowerId.Content.Ptr + length + sizeof(long)) = Bits.SwapBytes(GenerateNextEtag()); // now the id will be unique
+            
+            return newLowerId;
+        }
+
+        private const long ConflictedTombstoneIdMarkerLong = 20L;
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentsStorage.cs]: Why do you take an 8 bytes value for this?

Comment (ayende) [reply src/Raven.Server/Documents/DocumentsStorage.cs]: We typically use `30`, a byte, the record separator in ASCII

File: src/Raven.Server/Documents/DocumentsStorage.cs (hunk 2 above)
Comment (ayende) [inline src/Raven.Server/Documents/DocumentsStorage.cs]: Here you would check if the value is > 9, and if so, if lowerId[len-9] == 30

Comment (ayende) [reply src/Raven.Server/Documents/DocumentsStorage.cs]: However, you seems to be checking doing string comparisons here on value that isn't actually a string.

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Documents/DocumentsStorage.cs [hunk 4]
```diff
@@ -1556,14 +1561,45 @@ public void EnsureLastEtagIsPersisted(DocumentsOperationContext context, long do
...
+            var length = lowerId.Content.Length;
+            Slice.From(context.Allocator, lowerId.Content.Ptr, length + ConflictedTombstoneOverhead, out var newLowerId); // TODO: dispose?
+
+            *(newLowerId.Content.Ptr + length) = SpecialChars.RecordSeparator;
+            *(long*)(newLowerId.Content.Ptr + length + sizeof(byte)) = Bits.SwapBytes(GenerateNextEtag()); // now the id will be unique
+            
+            return newLowerId;
+        }
+
+        private const int ConflictedTombstoneOverhead = sizeof(long) + sizeof(byte);
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/DocumentsStorage.cs:1609]: Please add a comment that long is etag and byte is separator char

**Review by arekpalinski — APPROVED** [review]

**Comment by arekpalinski** [comment]
@karmeli87 One more thing came up do my mind. How `DeleteTombstoneIfNeeded` will behave then?

Will a tombstone with suffix be deleted on document put?

File: src/Raven.Server/Documents/DocumentsStorage.cs [hunk 5]
```diff
@@ -1556,14 +1561,45 @@ public void EnsureLastEtagIsPersisted(DocumentsOperationContext context, long do
...
 
             return (newEtag, changeVector);
         }
+        private Slice ModifyLowerIdIfNeeded(DocumentsOperationContext context, Table table, Slice lowerId)
+        {
+            if (table.ReadByKey(lowerId, out _) == false)
+                return lowerId;
+
+            var length = lowerId.Content.Length;
+            Slice.From(context.Allocator, lowerId.Content.Ptr, length + ConflictedTombstoneOverhead, out var newLowerId); // TODO: dispose?
```
Comment (arekpalinski) [inline src/Raven.Server/Documents/DocumentsStorage.cs]: TODO ?

Comment (karmeli87) [reply src/Raven.Server/Documents/DocumentsStorage.cs]: fixed

**Comment by karmeli87** [comment]
@arekpalinski You are correct.
`DeleteTombstoneIfNeeded` will delete only the tombstone of the relevant collection

File: src/Raven.Server/Documents/DocumentsStorage.cs [hunk 6]
```diff
@@ -828,11 +830,22 @@ public DocumentOrTombstone GetDocumentOrTombstone(DocumentsOperationContext cont
...
+                        Tombstone = TableValueToTombstone(context, ref item.Value.Reader)
+                    };
+                }
+                break;
+            }
 
             return new DocumentOrTombstone
             {
-                Tombstone = TableValueToTombstone(context, ref tvr)
+                Tombstone = null
```
Comment (ppekrol) [inline src/Raven.Server/Documents/DocumentsStorage.cs:849]: null?

File: src/Raven.Server/Documents/DocumentsStorage.cs (hunk 1 above)
Comment (ppekrol) [reply src/Raven.Server/Documents/DocumentsStorage.cs:1224]: How this affects all GetTombstonesFrom methods? E.g. What are we expecting to export via Smuggler? Let's talk about this on Sunday/Monday.

File: src/Raven.Server/Documents/DocumentPutAction.cs [hunk 7]
```diff
@@ -599,12 +599,16 @@ private static void DeleteTombstoneIfNeeded(DocumentsOperationContext context, C
             var tombstoneTable = context.Transaction.InnerTransaction.OpenTable(TombstonesSchema, collectionName.GetTableName(CollectionTableType.Tombstones));
             using (Slice.External(context.Allocator, lowerId, lowerSize, out Slice id))
             {
-                if (tombstoneTable.ReadByKey(id, out var reader) == false)
-                    return;
-
-                if (tombstoneTable.IsOwned(reader.Id))
+                foreach (var (tombstoneKey, tvh) in tombstoneTable.SeekByPrimaryKeyPrefix(id, Slices.Empty, 0))
```
Comment (ayende) [inline src/Raven.Server/Documents/DocumentPutAction.cs:602]: This will delete the existing values in the same collection, right?
But what about removing the postfix for the value when we delete the oldest value?

Comment (karmeli87) [reply src/Raven.Server/Documents/DocumentPutAction.cs:602]: Yes, this will delete the tombstone only if it in the given collection.

We don't care about removing the suffix, it is transparent to the other components since we always unwrap it (for replication, indexing, etl, etc.).

File: src/Raven.Server/Documents/DocumentsStorage.cs [hunk 8]
```diff
@@ -1608,7 +1609,7 @@ private static LazyStringValue UnwrapLowerIdIfNeeded(JsonOperationContext contex
 
             if (lowerId[lowerId.Size - ConflictedTombstoneOverhead] == SpecialChars.RecordSeparator)
             {
-                return new LazyStringValue(null, lowerId.Buffer, lowerId.Size - ConflictedTombstoneOverhead, context);
+                return context.GetLazyStringValue(lowerId.Buffer);
```
Comment (karmeli87) [inline src/Raven.Server/Documents/DocumentsStorage.cs]: @arekpalinski I do need to parse it for the export/backup, so I ended up doing it like this.

Comment (ppekrol) [reply src/Raven.Server/Documents/DocumentsStorage.cs]: Are we unwrapping here anything now? It looks like it takes the whole buffer so it will be the same as `lowerId`?

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs [hunk 9]
```diff
@@ -114,7 +114,7 @@ public long GetMinimalEtagForReplication()
                     foreach (var external in externals)
                     {
                         var state = GetExternalReplicationState(_server, Database.Name, external.TaskId, ctx);
-                        var myEtag = ChangeVectorUtils.GetEtagById(state.SourceChangeVector, Database.DbBase64Id);
+                        var myEtag = ChangeVectorUtils.GetEtagById(state.DestinationChangeVector, Database.DbBase64Id);
```
Comment (ayende) [inline src/Raven.Server/Documents/Replication/ReplicationLoader.cs:117]: why this change?

**Review by ayende — APPROVED** [review]

File: src/Raven.Server/Documents/Replication/ReplicationLoader.cs (hunk 9 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/Replication/ReplicationLoader.cs:117]: The "source" change vector is actually `_lastSentChangeVectorDuringHeartbeat`, which normally will be delayed in comparison to the destination change vector, which is the `LastAcceptedChangeVector` that gets updated every replication batch.

**Review by ppekrol — APPROVED** [review]
